libraries/rush-lib/src/logic/operations/NodeDiagnosticDirPlugin.ts

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import path from 'path';
+
+import { FileSystem } from '@rushstack/node-core-library';
+
+import type { IPhasedCommandPlugin, PhasedCommandHooks } from '../../pluginFramework/PhasedCommandHooks';
+import type { IEnvironment } from '../../utilities/Utilities';
+import type { Operation } from './Operation';
+import type { IOperationRunnerContext } from './IOperationRunner';
+import type { IOperationExecutionResult } from './IOperationExecutionResult';
+
+const PLUGIN_NAME: 'NodeDiagnosticDirPlugin' = 'NodeDiagnosticDirPlugin';
+
+export interface INodeDiagnosticDirPluginOptions {
+  diagnosticDir: string;
+}
+
+/**
+ * Phased command plugin that configures the NodeJS --diagnostic-dir option to contain the project and phase name.
+ */
+export class NodeDiagnosticDirPlugin implements IPhasedCommandPlugin {
+  private readonly _diagnosticsDir: string;
+
+  public constructor(options: INodeDiagnosticDirPluginOptions) {
+    this._diagnosticsDir = options.diagnosticDir;
+  }
+
+  public apply(hooks: PhasedCommandHooks): void {
+    const getDiagnosticDir = (operation: Operation): string | undefined => {
+      const { associatedProject } = operation;
+
+      if (!associatedProject) {
+        return;
+      }
+
+      const diagnosticDir: string = path.resolve(
+        this._diagnosticsDir,
+        associatedProject.packageName,
+        operation.logFilenameIdentifier
+      );
+
+      return diagnosticDir;
+    };
+
+    hooks.beforeExecuteOperation.tap(
+      PLUGIN_NAME,
+      (operation: IOperationRunnerContext & IOperationExecutionResult): undefined => {
+        const diagnosticDir: string | undefined = getDiagnosticDir(operation.operation);
+        if (!diagnosticDir) {
+          return;
+        }
+
+        // Not all versions of NodeJS create the directory, so ensure it exists:
+        FileSystem.ensureFolder(diagnosticDir);
+      }
```

Comment:
Why do this in the `beforeExecuteOperation` hook instead of `createEnvironmentForOperation` hook?

Reply:
Execution of `createEnvironmentForOperation` happens at the very beginning of the run. I wanted to spread out the cost of creating all of the diagnostics folders and attribute its cost to the individual operations.